File: dotnet/src/Core/History.cs
using System.Collections.Concurrent;
using QuikGraph;

namespace Agience.SDK
{
    public class InformationVertex
    {
        public string? Id { get; internal set; }
        public Data? Input { get; internal set; }
        public DateTime? InputTimestamp { get; internal set; }
        public Data? Output { get; internal set; }
        public DateTime? OutputTimestamp { get; internal set; }
        public string? FunctionId { get; internal set; }
    }

    public class InformationEdge : Edge<InformationVertex>
    {
        internal InformationEdge(InformationVertex source, InformationVertex target) : base(source, target) { }
    }

    public class History : BidirectionalGraph<InformationVertex, InformationEdge>
    {
        public string? Id { get; internal set; }
        public string? OwnerId { get; internal set; }

        private readonly ConcurrentDictionary<string, object> _vertexLocks = new ConcurrentDictionary<string, object>();
        private readonly ReaderWriterLockSlim _lock = new ReaderWriterLockSlim();

        public History(string? id = null, string? ownerId = null)
        {
            Id = id;
            OwnerId = ownerId;
        }

        internal void Add(Information information)
        {
            InformationVertex? currentVertex;

            var currentVertexLock = _vertexLocks.GetOrAdd(information.Id, _ => new object());

            lock (currentVertexLock)
            {
                _lock.EnterWriteLock();
                try
                {
                    // TODO: Use indexed dictionary for performance
                    currentVertex = Vertices.FirstOrDefault(v => v.Id == information.Id);

                    if (currentVertex == null)
                    {
                        currentVertex = new InformationVertex()
                        {
                            Id = information.Id,
                            Input = information.Input,
                            InputTimestamp = DateTime.TryParse(information.InputTimestamp!, out var inputTimestamp) ? inputTimestamp : null,
                            Output = information.Output,
                            OutputTimestamp = DateTime.TryParse(information.OutputTimestamp!, out var outputTimestamp) ? outputTimestamp : null,
                            FunctionId = information.FunctionId,
                        };
                        AddVertex(currentVertex);
                    }
                    else // Information builds in stages, but once the value is there it should be immutable. Safe to overwrite.
                    {
                        currentVertex.Input = information.Input;
                        currentVertex.InputTimestamp = DateTime.TryParse(information.InputTimestamp!, out var inputTimestamp) ? inputTimestamp : null;
                        currentVertex.Output = information.Output;
                        currentVertex.OutputTimestamp = DateTime.TryParse(information.OutputTimestamp!, out var outputTimestamp) ? outputTimestamp : null;
                        currentVertex.FunctionId = information.FunctionId;
                    }
                }
                finally
                {
                    _lock.ExitWriteLock();
                }
            }

            if (string.IsNullOrEmpty(information.ParentId)) { return; }

            var parentVertexLock = _vertexLocks.GetOrAdd(information.ParentId, _ => new object());

            lock (parentVertexLock)
            {
                _lock.EnterWriteLock();
                try
                {
                    // TODO: Use indexed dictionary for performance
                    var parentVertex = Vertices.FirstOrDefault(v => v.Id == information.ParentId);

                    if (parentVertex == null)
                    {
                        parentVertex = new InformationVertex()
                        {
                            Id = information.ParentId
                        };
                        AddVertex(parentVertex);
                    }

                    AddEdge(new InformationEdge(parentVertex, currentVertex));

                }
                finally
                {
                    _lock.ExitWriteLock();
                }
            }
        }
    }
}

File: core/sdk/python/history.py
import concurrent.futures
import datetime
from typing import Optional, Dict
from dataclasses import dataclass, field
from collections import defaultdict

@dataclass
class InformationVertex:
   id: Optional[str] = None
   input: Optional[dict] = None
   input_timestamp: Optional[datetime.datetime] = None
   output: Optional[dict] = None
   output_timestamp: Optional[datetime.datetime] = None
   function_id: Optional[str] = None

@dataclass
class InformationEdge:
   source: InformationVertex
   target: InformationVertex

class History:
   def __init__(self, id: Optional[str] = None, owner_id: Optional[str] = None):
       self.id = id
       self.owner_id = owner_id
       self.vertices: Dict[str, InformationVertex] = {}
       self.edges: Dict[str, InformationEdge] = {}
       self.vertex_locks: Dict[str, object] = defaultdict(object)
       self.lock = concurrent.futures.RLock()

   def add(self, information):
       with self.vertex_locks[information.id], self.lock:
           if information.id not in self.vertices:
               self.vertices[information.id] = InformationVertex(
                   id=information.id,
                   input=information.input,
                   input_timestamp=datetime.datetime.fromisoformat(information.input_timestamp) if information.input_timestamp else None,
                   output=information.output,
                   output_timestamp=datetime.datetime.fromisoformat(information.output_timestamp) if information.output_timestamp else None,
                   function_id=information.function_id,
               )
           else:
               self.vertices[information.id].input = information.input
               self.vertices[information.id].input_timestamp = datetime.datetime.fromisoformat(information.input_timestamp) if information.input_timestamp else None
               self.vertices[information.id].output = information.output
               self.vertices[information.id].output_timestamp = datetime.datetime.fromisoformat(information.output_timestamp) if information.output_timestamp else None
               self.vertices[information.id].function_id = information.function_id

       if information.parent_id:
           with self.vertex_locks[information.parent_id], self.lock:
               if information.parent_id not in self.vertices:
                   self.vertices[information.parent_id] = InformationVertex(id=information.parent_id)
               self.edges[f"{information.parent_id}-{information.id}"] = InformationEdge(
                   source=self.vertices[information.parent_id],
                   target=self.vertices[information.id]
               )